code-gen-projects/java/code-gen-demo/src/test/java/org/example/CodeGenTest.java

```diff
@@ -3,24 +3,71 @@
 import org.junit.jupiter.api.Test;
 import static org.junit.jupiter.api.Assertions.*;
 import java.util.ArrayList;
+import com.amazon.ion.system.IonReaderBuilder;
+import com.amazon.ion.IonReader;
+import com.amazon.ion.system.IonTextWriterBuilder;
+import com.amazon.ion.IonWriter;
+import com.amazon.ion.IonSystem;
+import com.amazon.ion.system.IonSystemBuilder;
+import com.amazon.ion.IonLoader;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.FileInputStream;
+import java.io.File;
 
 class CodeGenTest {
-    // TODO: Add roundtrip tests after read-write APIs are supported for Java code generation
+    private static final IonSystem ionSystem = IonSystemBuilder.standard().build();
+    private static final IonLoader ionLoader = ionSystem.getLoader();
+
     @Test void getterAndSetterTestForStructWithFields() {
          ArrayList<String> a = new ArrayList<String>();
          a.add("foo");
          a.add("bar");
          a.add("baz");
-         StructWithFields s = new StructWithFields("hello", 12, new AnonymousType2(a));
+         StructWithFields s = new StructWithFields("hello", 12, new AnonymousType2(a), 10e2);
          assertEquals("hello", s.getA(), "s.getA() should return \"hello\"");
          assertEquals(12, s.getB(), "s.getB() should return `12`");
          assertEquals(3, s.getC().getValue().size(), "s.getC().getValue() should return ArrayList fo size 3");
+         assertEquals(10e2, s.getD(), "s.getD() should return `10e2`");
     }
 
     @Test void getterAndSetterTestForNestedStruct() {
-             NestedStruct n = new NestedStruct("hello", 12, new AnonymousType1(false));
-             assertEquals("hello", n.getA(), "n.getA() should return \"hello\"");
-             assertEquals(12, n.getB(), "n.getB() should return `12`");
-             assertEquals(false, n.getC().getD(), "n.getC().getD() should return `false`");
+         NestedStruct n = new NestedStruct("hello", 12, new AnonymousType1(false));
+         assertEquals("hello", n.getA(), "n.getA() should return \"hello\"");
+         assertEquals(12, n.getB(), "n.getB() should return `12`");
+         assertEquals(false, n.getC().getD(), "n.getC().getD() should return `false`");
+    }
+
+    @Test void roundtripTestForStructWithFields() throws IOException {
```

Comment:
We should have tests for edge cases and failure cases to make sure the generated code handles those appropriately. Here are _some_ cases to consider.

Good—has an empty list:
```
{ A: "hello", B: 12, C: [], D: 10e2 }
```
Good—fields appearing in a different order:
```
{ D: 10e2, C: ["foo"], B: 12, A: "hello" }
```
Good(?)—optional field not present
```
{ B: 12, C: ["foo"], D: 10e2 }
```
Good(?)—sexp instead of a list for unspecified container type
```
{ A: "hello", B: 12, C: ("foo"), D: 10e2 }
```
Bad—wrong type for field:
```
{ A: "hello", B: 12, C: ["foo"], D: 123 }
```
Bad—wrong type for list elements:
```
{ A: "hello", B: 12, C: [true, false], D: 10e2 }
```
Bad—repeated field:
```
{ A: "hello", B: 12, C: ["foo"], D: 1e0, D: 1e0 }
```
Bad—null for non-null field:
```
{ A: null.string, B: 12, C: ["foo"], D: 1e0 }
```

Reply:
This makes sense but unfortunately there is nulls/typed nulls are not yet supported with this implementation. I will add other tests though.

Reply:
Also, current implementation doesn't process `occurs` for fields yet. Reference #102

Reply:
Add support for SExpressions: https://github.com/amazon-ion/ion-cli/issues/105
Add support for optional and repeated fields: https://github.com/amazon-ion/ion-cli/issues/106